monitor/supply.py

```diff
@@ -0,0 +1,110 @@
+import math
+from collections import OrderedDict
+from typing import Union, Optional, Dict
+
+import maya
+from maya import MayaDT
+from nucypher.blockchain.economics import BaseEconomics
+from nucypher.blockchain.eth.token import NU
+
+INITIAL_SUPPLY = NU(1_000_000_000, 'NU')
+UNIVERSITY_SUPPLY = NU(19_500_000, 'NU')
+CASI_SUPPLY = NU(9_000_000, 'NU')
+
+SAFT1_ALLOCATION_PERCENTAGE = 0.319
+SAFT2_ALLOCATION_PERCENTAGE = 0.08
+TEAM_ALLOCATION_PERCENTAGE = 0.106
+NUCO_ALLOCATION_PERCENTAGE = 0.2
+
+NUCO_VESTING_MONTHS = 5 * 12
+WORKLOCK_VESTING_MONTHS = 6
+UNIVERSITY_VESTING_MONTHS = 3 * 12
+
+LAUNCH_DATE = MayaDT.from_rfc3339('2020-10-15T00:00:00.0Z')
+DAYS_PER_MONTH = 30.416  # value used in allocations
+
+
+def vesting_remaining_factor(vesting_months: int,
+                             cliff: bool = False,
+                             now: Optional[MayaDT] = None) -> Union[float, int]:
+    """
+    Calculates the remaining percentage of tokens that should still be locked relative to launch date,
+    Oct 15, 2020 00:00:00 UTC, based on the provided vesting characteristics.
+    """
+    if not now:
+        now = maya.now()
+
+    months_transpired = math.floor((now - LAUNCH_DATE).days / DAYS_PER_MONTH)  # round down
+    if cliff:
+        return 1 if months_transpired < vesting_months else 0
+    else:
+        if months_transpired >= vesting_months:
+            # vesting period fully completed
+            return 0
+        else:
+            return (vesting_months - months_transpired) / vesting_months
+
+
+def calculate_supply_information(economics: BaseEconomics) -> Dict:
+    """Calculates the NU token supply information."""
+    supply_info = OrderedDict()
+    max_supply = NU.from_nunits(economics.total_supply)
+
+    # Initial Supply Information
+    initial_supply_info = OrderedDict()
+    supply_info['initial_supply'] = initial_supply_info
+    initial_supply_info['total_allocated'] = str(round(INITIAL_SUPPLY, 2))
+
+    # - Locked allocations
+    locked_allocations = OrderedDict()
+    initial_supply_info['locked_allocations'] = locked_allocations
+    now = maya.now()
+    vest_24_month_factor = vesting_remaining_factor(vesting_months=24, cliff=False, now=now)
+    vest_worklock_factor = vesting_remaining_factor(vesting_months=WORKLOCK_VESTING_MONTHS, cliff=True, now=now)
+    vest_nuco_factor = vesting_remaining_factor(vesting_months=NUCO_VESTING_MONTHS, cliff=True, now=now)
+    vest_university_factor = vesting_remaining_factor(vesting_months=UNIVERSITY_VESTING_MONTHS, cliff=True, now=now)
```

Comment:
@mswilkison pay particular attention here for vesting calcs, and how they are used.